### DISCUSSION TITLE
A case for module-level variables

### DISCUSSION THREAD

**Discussion by srackham** [vlang/v#16367] [Ideas]
Module-level variables with the optional `pub` accessor would be a decided improvement over `__global`.

`__global` variables are application-wide (they break module encapsulation) and force you to use the perjorative `-enable-globals` compiler option globally. If a dependency requires `-enable-globals` then so do all client applications that use the dependency (V has implemented a `[has_globals]` tag to turn off `-enable-globals` but it's only available inside the V codebase).

Top-level variables are not prohibited in V scripts and they're not memory-unsafe.

Maybe module variables are "bad practice" but it's just one of many anti-patterns.

Porting existing code, for example from Go (which allows package-level variables), should not force an unnecessary rewrite (see [Things You Should Never Do](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/)).

Yes, you could implement singletons (convoluted in V) but singleton encapsulation is little more than syntatic sugar plus additional congitive and calling overheads.

See also #13926.

**Comment by dumblob** [comment]
It is very difficult to find a simple and **guaranteed** semantics for anything module-level in runtime due to parallelism, scheduling, lifetimes etc. Please read https://github.com/vlang/v/issues/9388 (and https://github.com/vlang/v/discussions/11835 ).

All in all I am not in favor of anything module-level in runtime.

**Reply by JalonSolov** [reply]
These would just be globals with a smaller scope.

Sure, they would be convenient for some things (like porting bad code from Go, Perl, and other languages that support this idea), but they also conveniently carry the same problems.

**Comment by srackham** [comment]
I'm not proposing module `init`/`deinit` functions or really anything new, just that globals are made easier and more natural to use by:

1. Module-level namespacing: private by default, public with `pub`.

2. Drop the neccessity for the `-enable-globals` flag for module-level variables.

3. Drop the ugly pejorative `__global` keyword or replace it with `var` (cf. Go).

I do understand that globals are open to abuse especially with concurrency, but they are useful and they are easy to understand and use. They're not intrinsically unsafe and they hardly even raise an eyebrow in other mainstream languages (Go, Python, Kotlin, Dart, TypeScript, JavaScript ...).

The compiler has a role to advise with optional warnings, but not to arbitrate morals by making things unnecessarily difficult.

**Reply by JalonSolov** [reply]
They are definitely unsafe in a multi-tasking environment.  The "perjorative" `__global` is to make sure you know what you're doing (and the risks you're taking) by using it.

**Reply by medvednikov** [reply]
Exactly :)

**Comment by srackham** [comment]
@JalonSolov By "intrinsically unsafe" I meant not unavoidably unsafe and not memory unsafe in the same way that segfaulting pointer arithmetic is. Concurrency data races are avoided by the use of mutex locks. Forbidding globals won't put an end to data races, if that's the goal then you'll also need a compiler `-enable-spawn` option and rename `spawn` to `__spawn`.

How does the use of `__global` make the user know what they're doing any more than a compiler warning?

Personally, I think that these sorts of subjective calls should be left to separate lint-style sanity checker commands -- the compilers job is to  deal with broken syntax (a job it does well) not broken semantics (a job it will always do badly).

**Reply by dumblob** [reply]
As discussed in the linked issue(s) above https://github.com/vlang/v/issues/9388 - the very first thing is how to make it **fully explicit** how many instances of a "module global variable" would you have in runtime. Any ideas?

`__global` solves this.

>How does the use of `__global` make the user know what they're doing any more than a compiler warning?

That is what *V* understands under **explicit** - namely **reading** the code describes the whole picture without reliance on the compiler and its analysis. Compiler warning is thus **not** explicit.

`spawn` is explicit. IMHO it is not `__spawn` because spawn is essential to access other *V* features - i.e. is a dependency for the language features themself (in this case it is multithreading and friends). `__global` is not essential (actually not needed at all if I am not mistaken) for *V* features. In other words nothing in *V* depends on `__global` (if this is not the case then it is probably a bug - please file a report).

**Reply by srackham** [reply]
Thanks for your patient explanation, I'm starting to get it but don't have an answer (this sort of stuff is above my pay grade).

All I know is that I have a library that I've ported from TypeScript to Go, Kotlin, Python and Dart without any issues (all five languages handle module-level variables with ease), but with V all module static variables have to be broken out of encapsulation and any apps that want to use the library (directly or indirectly) have to be compiled with the `-enable-globals` option.

I suppose I could use singletons but that just seems (correct me if I'm wrong) to be nothing more than sleight of hand to trick the compiler out of having to use `-enable-globals` -- an unnecessary complication that makes a verbose mess of all the reference sites. I've been unable to come up with a more elegant solution that doesn't involve rearchitecting the codebase.

FWIW here's the unsafe singleton implementation I was playing with (it stores the data in a function static variable):

```v
module singleton

/*
The singleton should be a struct with the data residing in struct fields.
See https://discordapp.com/channels/592103645835821068/592294828432424960/934198551884468275
*/

[unsafe]
fn private_get<T>() &T {
	mut static s := &T(0)
	if u64(s) == 0 {
		s = &T{}
	}
	return s
}

pub fn get<T>() &T {
	return unsafe { private_get<T>() }
}
```

**Reply by dumblob** [reply]
@srackham I see and feel your pain. This is a perfect demonstrative example of the subtle but absolutely necessary difference between the languages.

I can only confirm (and am very happy about this!) that *V*'s design have pushed you to made it **explicit** and difficult for you to achieve something unsafe (unsafe in terms of "error-prone for humans" due to lack of deep human understanding of the subtle implicit semantics of languages as Go, Kotlin, Python, Dart).

What you want to achieve is deliberately verbose (`static` is not recommended) if not rearchitectured (as you indicated). Yes, *V* is kind of a middle ground between Rust and other conventional languages (Go, Kotlin, Dart) in that it sometimes requires you to adhere to certain (believed to be better) architecture patterns instead of trying to keep the (worse) ones in place. Yes, it complicates porting of existing apps/libs but the *V*'s semantics is built around "there is only one way of doing a thing" policy which as we - people around *V* - believe is truly helpful in the end.

This probably did not make you happy in a short run - though I am absolutely certain this very same thing will make you happy in the long run :wink:.

**Reply by srackham** [reply]
Thanks for that, I do like V's "one way" policy and I do except that there are considerations that make module globals technically difficult to implement in V.

Having ported my code the problem is now the mandatory `-enable-globals` option that all module users are forced to build with -- it screams "danger, unsafe code ahead!".  Heck, the V developers have even implemented a `[has_globals]` tag to work around `-enable-globals`, but it's only available inside the V codebase. Why not make this tag available to everyone? Surely that's not a technical issue.

I've thought long and hard about rearchitecting the code using various design patterns, but that would be a huge time-sink and add nothing except complexity and obfustication (https://wiki.c2.com/?DesignPatternsConsideredHarmful). Plus the port would be out of sync with the other ports which would complicate maintenance. Hard to see how this would make me "happy in the long run" :-)

The code in question was not casually thrown together, it does what it was designed to do and is not unsafe. It passes a battery of around 300 identical tests in all languages, many of the tests winkled out with extensive Go fuzz testing.

**Reply by dumblob** [reply]
>Why not make this tag available to everyone? Surely that's not a technical issue.

The idea is to support the theory that in a safe-by-default language there is no **need** for this.

>I've thought long and hard about rearchitecting the code using various design patterns, but that would be a huge time-sink and add nothing except complexity and obfustication (https://wiki.c2.com/?DesignPatternsConsideredHarmful). Plus the port would be out of sync with the other ports which would complicate maintenance. Hard to see how this would make me "happy in the long run" :-)

I think this is more of a philosophical question. If one wants to translate things between languages and expects extreme similarity of the translated stuff, then maybe those languages are very close to each other in which case I can only say "V is different - and for good reason".

So it is probably more about the acceptance that things can (and maybe should? - only future will show...) be done differently. I as a latin-alphabet native would never expect my poetry to be identically or even similarly translated to mandarin. Because I accept mandarin is different - and for good reason.

So, why do you actually want to translate the SW without accommodating it to *V* (read "rearchitecture it") instead of proxying it through a high-level interface based on *V*'s good practices (even if you ended up using 0-overhead FFI-like mechanisms - which V supports - under the hood)?

>The code in question was not casually thrown together, it does what it was designed to do and is not unsafe. It passes a battery of around 300 identical tests in all languages, many of the tests winkled out with extensive Go fuzz testing.

This further supports the approach to actually **not** "translate the source" but rather only interface with the *code* using *V* semantics instead of the originally meant semantics.

But maybe I am missing something - feel free to ask even on discord (where I am not available as that would eat up the very few last minutes I have).

**Reply by srackham** [reply]
That's a great reply. I take your point that the real problem is my trying to fit a square peg into a round hole (I do these ports to learn new languages and to see how the language fits the code rather than vice versa). V is very nice, it's the Go we'd get if we could start all over again -- if native compilation pans out it will be dynamite.

Back to my problem: I think we're left at a philosopical impasse and that's ok, languages are more than just technical vehicles.
